turf/src/commonMain/kotlin/org/maplibre/spatialk/turf/other/RTree.kt

```diff
@@ -0,0 +1,712 @@
+import org.maplibre.spatialk.geojson.LineString
+import org.maplibre.spatialk.geojson.MultiLineString
+import org.maplibre.spatialk.geojson.MultiPoint
+import org.maplibre.spatialk.geojson.MultiPolygon
+import org.maplibre.spatialk.geojson.Point
+import org.maplibre.spatialk.geojson.Polygon
+import org.maplibre.spatialk.turf.coordinatemutation.flattenCoordinates
+import org.maplibre.spatialk.turf.measurement.computeBbox
+
+/**
+ * A high-performance R-tree-based spatial index for 2D points and rectangles.
+ *
+ * RTree is an implementation of an R-tree, a tree data structure used for spatial access methods,
+ * i.e., for indexing multi-dimensional information such as geographical coordinates, rectangles, or
+ * polygons. It's a popular choice for spatial indexing due to its efficiency in handling dynamic
+ * datasets (i.e., data that changes over time) and its speed in performing spatial queries.
+ *
+ * This implementation is a port of the JavaScript `RBush` library and is optimized for indexing and
+ * querying large datasets of geographical features. It supports operations like insertion, removal,
+ * and spatial searches (e.g., finding all items within a given bounding box).
+ *
+ * The performance of the R-tree is heavily influenced by the `maxEntries` parameter. A higher value
+ * for `maxEntries` leads to a tree with fewer, fuller nodes. This can speed up initial loading and
+ * insertion operations but may slow down search queries because each node covers a larger area.
+ * Conversely, a lower `maxEntries` value creates a more fine-grained tree with smaller nodes, which
+ * can be faster for searches but slower for insertions. The default value of 16 is a commonly used
+ * and well-balanced choice.
+ *
+ * @param T The type of [Feature] to be stored in the tree. The feature's geometry is used to
+ *   calculate its bounding box for indexing.
+ * @param maxEntries The maximum number of entries in a single tree node. A higher value leads to
+ *   faster loading/insertion but slower searches, and vice versa. Must be 4 or greater. Defaults
+ *   to 16.
+ */
+public class RTree<T : GeoJsonObject>(initialData: List<T> = emptyList(), maxEntries: Int = 16) {
+    internal data class Node<T : GeoJsonObject>(
+        var item: T? = null,
+        var children: MutableList<Node<T>> = mutableListOf(),
+        var height: Int = 1,
+        var leaf: Boolean = true,
+        var bbox: BoundingBox =
+            item?.bbox
+                ?: BoundingBox(
+                    Double.POSITIVE_INFINITY,
+                    Double.POSITIVE_INFINITY,
+                    Double.NEGATIVE_INFINITY,
+                    Double.NEGATIVE_INFINITY,
+                ),
+    ) {
+        fun computeBbox() {
+            bbox =
+                if (item != null) {
+                    item?.bbox!!
+                } else {
+                    computeBbox(children.flatMap { listOf(it.bbox.northeast, it.bbox.southwest) })
```

Comment:
would using `buildList` here improve performance by avoiding all the sub-list and vararg array allocations?